src/util/submit-addon.js

```diff
@@ -0,0 +1,344 @@
+/* @flow */
+import { createWriteStream, promises as fsPromises } from 'fs';
+import { pipeline } from 'stream';
+import { promisify } from 'util';
+
+// eslint-disable-next-line no-shadow
+import fetch, { FormData, fileFromSync, Response } from 'node-fetch';
+import { SignJWT } from 'jose';
+
+import {createLogger} from './../util/logger.js';
+
+const log = createLogger(import.meta.url);
+
+export type SignResult = {|
+  id: string,
+  downloadedFiles: Array<string>,
+|};
+
+type ClientConstructorParams = {|
+  apiKey: string,
+  apiSecret: string,
+  apiHost: string,
+  apiJwtExpiresIn?: number,
+  validationCheckInterval?: number,
+  validationCheckTimeout?: number,
+  approvalCheckInterval?: number,
+  approvalCheckTimeout?: number,
+  downloadDir?: string,
+|};
+
+export default class Client {
+  apiKey: string;
+  apiSecret: string;
+  apiUrl: string;
+  apiJwtExpiresIn: number;
+  validationCheckInterval: number;
+  validationCheckTimeout: number;
+  approvalCheckInterval: number;
+  approvalCheckTimeout: number;
+  downloadDir: string;
+
+  constructor({
+    apiKey,
+    apiSecret,
+    apiHost,
+    apiJwtExpiresIn = 60 * 5, // 5 minutes
+    validationCheckInterval = 1000,
+    validationCheckTimeout = 300000, // 5 minutes.
+    approvalCheckInterval = 1000,
+    approvalCheckTimeout = 900000, // 15 minutes.
+    downloadDir = process.cwd(),
+  }: ClientConstructorParams) {
+    this.apiKey = apiKey;
+    this.apiSecret = apiSecret;
+    this.apiUrl = `${apiHost}/api/v5/addons/`;
```

Comment:
Sorry, I'm not following this comment at all 😅 

The hard-coded `v5/api/` prefix with an `apiHost` was a requested change a few weeks back - look back in the comments on this PR.  Originally an Errror was thrown with almost exactly that wording!

I'm also not sure the relevance of using `URL` rather than strings is to hardcoding urls?

Reply:
see https://github.com/mozilla/web-ext/pull/2458#discussion_r930841340 and the other comments from the same review.

Reply:
be able to find a resolved comment in a github PR is easier said than done :-P

But I think I found the comment where the previous discussion about this happened, it should be https://github.com/mozilla/web-ext/pull/2458#discussion_r930841340 right?

I think I totally missed that mention at that time.

Personally I'm not super thrilled to have the `apiUrl` prefix hardcoded, but not really that big of a deal, especially while the new implementation is still behind an explicit command line option that needs to be passed to activate it, I'm fine to leave it as is for now (also given that William is currently away, and I'd prefer to gather some more from his perspective before asking to rework something already agreed as part of William review passes), we can still reconsider that as a separate followup.

The `URL` part of my comment was related but not completely overlapping with the "hardcoded path prefix" part of my comment, I'm sorry if that was more confusing than helpful.

In short I don't like that much how we are deriving the host and scheme part from the `apiHost` cli option value, in particular where the current version of this patch is currently doing the following:
```
      if (useSubmissionApi && apiHost.endsWith('/')) {
        apiHost = apiHost.slice(0, -1);
      }
```

That is kind of assuming that the user would be passing header "https://addons.mozilla.org" or "https://addons.mozilla.org/", but in the end the user may as well pass something like "https://addons.mozilla.org/somepath/" or some other kind of url.

With `URL` available we could use something like:
```
        try {
          const apiHostURL = new URL(apiHost);
          apiHost = apiHostURL.origin
        } catch (err) {
          throw new UsageError(`Invalid apiHost: ${apiHost}`);
        }
```

which should be slightly less fragile than using `endsWith` + `.slice`, and if we want to be more strict we could even check that `apiHostURL` didn't include any unexpected component that we were not expecting (e.g. if there is a `pathname`, `username`, `password`, `search` or `hash` set to a non empty string).

Reply:
I added your snippet that uses `URL` to parse the url.  And also changed everywhere there is a url in `submit-addon` to use `URL` for the type.  (`apiHost` was a bit of an edge case, as it's not a full url, so I left it as a string)